DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/annotated_ptr/src/annotated_ptr.cpp

```diff
@@ -0,0 +1,137 @@
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/sycl.hpp>
+
+#include "exception_handler.hpp"
+
+#include <iomanip>
+
+using namespace sycl;
+using namespace ext::intel::experimental;
+using namespace ext::oneapi::experimental;
+using usm_buffer_location =
+    ext::intel::experimental::property::usm::buffer_location;
+
+constexpr int kBL1 = 1;
+constexpr int kBL2 = 2;
+using MyPipe = ext::intel::experimental::pipe<class MyPipeName, float *>;
+
+#define ROWS 2
+#define COLS 5
+
+// Launch a kernel that does a weighted sum over a matrix
+// result[0] = data[0][0]*mul[0] + data[0][1]*mul[1] + ... +
+// result[1] = data[1][0]*mul[0] + data[1][1]*mul[1] + ... +
+//          ...
+struct pipeWithAnnotatedPtr {
+  annotated_arg<float *, decltype(properties{buffer_location<kBL1>})> result;
+  annotated_arg<float *, decltype(properties{buffer_location<kBL2>})> mul;
+
+  void operator()() const {
+    for (int i = 0; i < ROWS; i++) {
+      float *p = MyPipe::read();
+
+      // set buffer location on p with annotated_ptr
+      annotated_ptr<float, decltype(properties{buffer_location<kBL1>})> data{p};
```

Comment:
```suggestion
      annotated_ptr<float, decltype(properties{buffer_location<kBL3>})> data{p};
```
Would it make sense to put `data` in a 3rd buffer location so that it's clear that the global memory accesses won't arbitrate? that will be a simpler data model (although the IP will have 3 mm host interfaces)

Reply:
No the buffer location of the pointers read from pipe has to be one `kBL1` of `kBL2`, because the kernel mmhost interfaces are determined by the annotations on the kernel arguments.  So specifying a local variable to `kBL3` will result in an error like "can't find global memory kBL3 in board spec"

Reply:
what if you add an extra kernel argument?

Reply:
This design (dot product between matrix and vector) mainly consists of 3 parts of data:
- input vector (kBL2, passed by kernel argument)
- output vector (kB1,  passed by kernel argument)
- matrix (kB1,  passed by host pipe and annotated inside the kernel)

The two kernel arguments are necessary in generating the two mmhosts, and the matrix annotated using `annotated_ptr` is the gist of this code sample. An extra kernel argument seems not necessary regarding the current functionality of the design.